<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>Index. iEDGE 1.0</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content="Amy Li">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["\\(","\\)"] ],
      processEscapes: true
    }
  });
</script>
<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class="container">
      <a class="brand" href="#">iEDGE 1.0</a>
      <div class="nav">
        <ul class="nav">
          <li><a href="index.html"><i class="icon-home icon-white"></i> Index</a></li>
        </ul>
      </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <div class="row">
  <div class="span8">
    <h1>iEDGE</h1>

<p><strong>I</strong>ntegrative analysis of (<strong>E</strong>pi-)<strong>D</strong>NA and <strong>G</strong>ene <strong>E</strong>xpression Data</p>

<p>iEDGE is a R-package for performing integrative analysis of epi-DNA and gene expression data. This analysis was based on <a href="http://www.ncbi.nlm.nih.gov/pubmed/22975378">S. Monti et. al. Cancer Cell (2012).</a>, which describes an approach to integrate copy number alteration to gene expression dataset in Diffuse Large B cell lymphoma samples and extends the methodology to study in detail the effect of p53 and cell cycle deregulation in DLBCL. This package provides the generalized pipeline to perform integrative analysis of epi-genomic data (methylation, copy number alteration, mutation, microRNA, etc) and gene-expression data arising from overlapping samples. The package consistent of 3 major modules:
1. cis/trans differential expression based analysis of epi-DNA and gene expression data
2. pathway enrichment analysis of significant cis/trans correlations
3. pruning of cis/trans epi-DNA to gene expression connections using conditional independence criteria</p>

<p>Reporting options are available in table (.txt) format as well as in html format, supported by interactive plots and tables</p>

<h1>Installing iEDGE</h1>

<p>(1) Please install the following Bioconductor packages first before installing iEDGE:</p>

<pre><code>source(&quot;http://bioconductor.org/biocLite.R&quot;)
biocLite(c(&quot;Biobase&quot;, &quot;limma&quot;, &quot;org.Hs.eg.db&quot;))

</code></pre>

<p>(2) Installing iEDGE</p>

<pre><code>
library(devtools)

PAT &lt;- &quot;04fe676593e46b6bda5a5d09431156e8a500349a&quot;

install_github(&quot;montilab/iEDGE&quot;,auth_token = PAT)

library(iEDGE)

</code></pre>


    <h2>Help topics</h2>

    <h3></h3>
    
    
    <ul class="index">
        
      <li>
        <code><a href="iEDGE_combine.html">iEDGE_combine</a></code><br />iEDGE_combine combines alteration data (e.g. GISTIC2) and gene expression data, using a mapping data frame for matching samples</li>
            
      <li>
        <code><a href="iEDGE_DE.html">iEDGE_DE</a></code><br />iEDGE_DE performs differential expression analysis and pathway enrichment and saves to text tables</li>
            
      <li>
        <code><a href="iEDGE_UI.html">iEDGE_UI</a></code><br />iEDGE_UI makes the user interface for iEDGE reports</li>
            
      <li>
        <code><a href="make_GISTIC2.html">make_GISTIC2</a></code><br />make_GISTIC2 preprocess GISTIC2 data</li>
            
      <li>
        <code><a href="read_gmt.html">read_gmt</a></code><br />read_gmt</li>
            
      <li>
        <code><a href="run_hyperEnrichment.html">run_hyperEnrichment</a></code><br />run_hyperEnrichment</li>
            
      <li>
        <code><a href="run_hyperEnrichment_wrapper.html">run_hyperEnrichment_wrapper</a></code><br />run_hyperEnrichment_wrapper</li>
            
      <li>
        <code><a href="run_iEDGE.html">run_iEDGE</a></code><br />run_iEDGE is the main wrapper for iEDGE, assumes processed data</li>
    
    </ul>
  </div>
  
  <div class="span3 offset1">
        
    <h2>Dependencies</h2>
    <ul>
      
      <li><strong>Imports</strong>: Biobase, limma, plyr, infotheo, XML, knitr, RColorBrewer, org.Hs.eg.db, AnnotationDbi, cba</li>
      
      
    </ul>
    <h2>Authors</h2>
    <ul>
    </ul>

  </div>
</div>
      
      <footer>
      <p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>