<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
<title>process_data. Gistic2GE 1.0</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content="">

<link href="css/bootstrap.css" rel="stylesheet">
<link href="css/bootstrap-responsive.css" rel="stylesheet">
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="navbar">
  <div class="navbar-inner">
    <div class="container">
      <a class="brand" href="#">Gistic2GE 1.0</a>
      <div class="nav">
        <ul class="nav">
          <li><a href="index.html"><i class="icon-home icon-white"></i> Index</a></li>
        </ul>
      </div>
    </div>
  </div>
</div>

    <div class="container">
      <header>
        
      </header>
      
      <h1><code>process_data</code> process gistic and ge data from text files</h1>

<div class="row">
  <div class="span8">
    <h2>Usage</h2>
    <pre><div>process_data(amp.peak.file, amp.region.file&nbsp;=&nbsp;NA, del.peak.file, del.region.file&nbsp;=&nbsp;NA, lesions.file, ge.file, fmap.file, ge.confounder.file&nbsp;=&nbsp;NA, binary&nbsp;=&nbsp;TRUE, match.by&nbsp;=&nbsp;c("Descriptor", "cytoband"), fmap.gistic.col&nbsp;=&nbsp;"snp.array.ID", fmap.ge.col&nbsp;=&nbsp;"mrna.array.ID", row.ind&nbsp;=&nbsp;FALSE)</div></pre>
    
    <h2>Arguments</h2>
    <dl>
      <dt>amp.peak.file</dt>
      <dd>GISTIC amp genes in peak file</dd>
      <dt>amp.region.file</dt>
      <dd>GISTIC amp genes in region file</dd>
      <dt>del.peak.file</dt>
      <dd>GISTIC del genes in peak file</dd>
      <dt>del.region.file</dt>
      <dd>GISTIC del genes in region file</dd>
      <dt>lesions.file</dt>
      <dd>GISTIC all_lesions file</dd>
      <dt>ge.file</dt>
      <dd>Gene Expression res file</dd>
      <dt>fmap.file</dt>
      <dd>file mapping tabular file from GISTIC to GE sample names</dd>
      <dt>ge.confounder.file</dt>
      <dd>confounder indicator cls file, must match order in fmap.file</dd>
      <dt>...</dt>
      <dd>see addition arguments passed to <code>match_samples, read_res, read_cis_genes, read_gistic</code></dd>
    </dl>
    
    <div class="Value">
      <h2>Value</h2>

      <p><dl>
list containing :
<dt>amp</dt><dd>amplification gistic and ge</dd></p>
  
      <p><dt>del</dt><dd>deletion gistic and ge</dd></p>
  
      <p></dl></p>
  
    </div>

    <div class="Description">
      <h2>Description</h2>

      <p><code>process_data</code> process gistic and ge data from text files</p>
  
    </div>
      </div>
  <div class="span4">
    <!-- <ul>
      <li>process_data</li>
    </ul>
    <ul>
      
    </ul> -->
      
        
  </div>
</div>
      
      <footer>
      <p class="pull-right"><a href="#">Back to top</a></p>
<p>Built by <a href="https://github.com/hadley/staticdocs">staticdocs</a>. Styled with <a href="http://twitter.github.com/bootstrap">bootstrap</a>.</p>
      </footer>
    </div>
  </body>
</html>