<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>iEDGE analysis of a simulated dataset • iEDGE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="iEDGE analysis of a simulated dataset">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">iEDGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/iedge-simple-run.html">iEDGE analysis of a simulated dataset</a>
    </li>
    <li>
      <a href="../articles/iedge-tcga-example.html">iEDGE analysis of the TCGA PAAD dataset</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/iEDGE">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>iEDGE analysis of a simulated dataset</h1>
                        <h4 class="author">Amy Li</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/iEDGE/blob/master/vignettes/iedge-simple-run.Rmd"><code>vignettes/iedge-simple-run.Rmd</code></a></small>
      <div class="hidden name"><code>iedge-simple-run.Rmd</code></div>

    </div>

    
    
<p>For this tutorial we will simulate the input gene expression data, the cis and trans genes with known mediations, and we will run iEDGE to recover these mediations.</p>
<div id="setting-up-the-iedge-object" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-iedge-object" class="anchor"></a>Setting up the iEDGE object</h2>
<p>We start by generating the dataset. You can use <code>construct_iEDGE</code> to generate the iEDGE dataset from standard matrices (binary matrices for alteration data or numeric matrices for expression data).</p>
<p>Please note that this simulation is optimized for the evaluation of the iEDGE methology, and its only purpose is to demonstrate how to construct an iEDGE object from scratch for use in standard analysis.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(iEDGE)</span>
<span id="cb1-2"><a href="#cb1-2"></a></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">1</span>)</span>
<span id="cb1-4"><a href="#cb1-4"></a></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co"># define number of samples</span></span>
<span id="cb1-6"><a href="#cb1-6"></a>n &lt;-<span class="st"> </span><span class="dv">50</span></span>
<span id="cb1-7"><a href="#cb1-7"></a></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co"># construct 'cn': binary alternations by samples matrix two</span></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="co"># independent alterations across 50 samples</span></span>
<span id="cb1-10"><a href="#cb1-10"></a>cn1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="dt">x=</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">0</span>), <span class="dt">size =</span> n, <span class="dt">replace =</span> T, <span class="dt">prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.7</span>, <span class="fl">0.3</span>))</span>
<span id="cb1-11"><a href="#cb1-11"></a>cn2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="dt">x=</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">0</span>), <span class="dt">size =</span> n, <span class="dt">replace =</span> T, <span class="dt">prob =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>))</span>
<span id="cb1-12"><a href="#cb1-12"></a></span>
<span id="cb1-13"><a href="#cb1-13"></a>cn &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(cn1, cn2)</span>
<span id="cb1-14"><a href="#cb1-14"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(cn) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"sample_"</span>, <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(cn))</span>
<span id="cb1-15"><a href="#cb1-15"></a></span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="co"># construct 'cis' gene expressions: genes correlated with one or more</span></span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="co"># alterations in 'cn'</span></span>
<span id="cb1-18"><a href="#cb1-18"></a>cis &lt;-<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>), <span class="cf">function</span>(i) {</span>
<span id="cb1-19"><a href="#cb1-19"></a>        <span class="dv">3</span> <span class="op">+</span><span class="st"> </span>cn[i, ] <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n, <span class="dv">0</span>, <span class="fl">0.3</span>)</span>
<span id="cb1-20"><a href="#cb1-20"></a>}))</span>
<span id="cb1-21"><a href="#cb1-21"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(cis) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/make.unique">make.unique</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"cis"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>)))</span>
<span id="cb1-22"><a href="#cb1-22"></a></span>
<span id="cb1-23"><a href="#cb1-23"></a><span class="co"># construct 'trans-mediated' gene expression: gene correlated with</span></span>
<span id="cb1-24"><a href="#cb1-24"></a><span class="co"># one or more alteration in 'cn' and mediated through a cis gene</span></span>
<span id="cb1-25"><a href="#cb1-25"></a></span>
<span id="cb1-26"><a href="#cb1-26"></a>cis.m &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(cis), <span class="dv">10</span>, <span class="dt">replace =</span> T)</span>
<span id="cb1-27"><a href="#cb1-27"></a>trans.m &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(cis.m), <span class="cf">function</span>(i)</span>
<span id="cb1-28"><a href="#cb1-28"></a>        <span class="dv">3</span> <span class="op">+</span><span class="st"> </span>cis[cis.m[i], ] <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n, <span class="dv">0</span>, <span class="fl">0.3</span>)</span>
<span id="cb1-29"><a href="#cb1-29"></a>    ))</span>
<span id="cb1-30"><a href="#cb1-30"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(trans.m) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/make.unique">make.unique</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"t.m."</span>, cis.m))</span>
<span id="cb1-31"><a href="#cb1-31"></a></span>
<span id="cb1-32"><a href="#cb1-32"></a><span class="co"># construct 'trans-direct' gene expression: trans gene correlated</span></span>
<span id="cb1-33"><a href="#cb1-33"></a><span class="co"># with one or more alteration in 'cn' but not mediated through cis</span></span>
<span id="cb1-34"><a href="#cb1-34"></a><span class="co"># genes, this would be similar to construction of cis genes</span></span>
<span id="cb1-35"><a href="#cb1-35"></a>trans.d &lt;-<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>), <span class="cf">function</span>(i)</span>
<span id="cb1-36"><a href="#cb1-36"></a>    <span class="dv">3</span> <span class="op">+</span><span class="st"> </span>cn[i, ] <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n, <span class="dv">0</span>, <span class="fl">0.6</span>)</span>
<span id="cb1-37"><a href="#cb1-37"></a>    ))</span>
<span id="cb1-38"><a href="#cb1-38"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(trans.d) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/make.unique">make.unique</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"t.d."</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>)))</span>
<span id="cb1-39"><a href="#cb1-39"></a></span>
<span id="cb1-40"><a href="#cb1-40"></a><span class="co"># uncorrelated genes</span></span>
<span id="cb1-41"><a href="#cb1-41"></a>others &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">30</span>, <span class="cf">function</span>(i) <span class="dv">3</span><span class="op">+</span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n, <span class="dv">0</span>, <span class="fl">0.3</span>)))</span>
<span id="cb1-42"><a href="#cb1-42"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(others) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"others_"</span>, <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(others))</span>
<span id="cb1-43"><a href="#cb1-43"></a></span>
<span id="cb1-44"><a href="#cb1-44"></a><span class="co"># putting genes together in expression matrix</span></span>
<span id="cb1-45"><a href="#cb1-45"></a>gep &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(cis, trans.m, trans.d, others)</span>
<span id="cb1-46"><a href="#cb1-46"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(gep) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"sample_"</span>, <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(gep))</span>
<span id="cb1-47"><a href="#cb1-47"></a></span>
<span id="cb1-48"><a href="#cb1-48"></a>cisgenes &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">cn1 =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"cis1"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(cis), <span class="dt">value =</span> T),</span>
<span id="cb1-49"><a href="#cb1-49"></a>                 <span class="dt">cn2 =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">grep</a></span>(<span class="st">"cis2"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(cis), <span class="dt">value =</span> T))</span>
<span id="cb1-50"><a href="#cb1-50"></a></span>
<span id="cb1-51"><a href="#cb1-51"></a><span class="co"># the direction column in cn.fdat specifies whether to compute</span></span>
<span id="cb1-52"><a href="#cb1-52"></a><span class="co"># one-sided or two-sided p-values in differential tests. In this</span></span>
<span id="cb1-53"><a href="#cb1-53"></a><span class="co"># simulation we expect an increase in expression with the presence of</span></span>
<span id="cb1-54"><a href="#cb1-54"></a><span class="co"># the alteration, hence a one-sided ("UP") p-value is computed.</span></span>
<span id="cb1-55"><a href="#cb1-55"></a>cn.fdat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">cnid =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(cn), <span class="dt">direction =</span> <span class="st">"UP"</span>)</span>
<span id="cb1-56"><a href="#cb1-56"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(cn.fdat) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(cn)</span>
<span id="cb1-57"><a href="#cb1-57"></a></span>
<span id="cb1-58"><a href="#cb1-58"></a>gep.fdat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="dt">geneid =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(gep))</span>
<span id="cb1-59"><a href="#cb1-59"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(gep.fdat) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(gep)</span>
<span id="cb1-60"><a href="#cb1-60"></a></span>
<span id="cb1-61"><a href="#cb1-61"></a>dat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/construct_iEDGE.html">construct_iEDGE</a></span>(cn, gep, cisgenes, <span class="dt">transgenes =</span> <span class="ot">NA</span>, cn.fdat, gep.fdat, </span>
<span id="cb1-62"><a href="#cb1-62"></a>                <span class="dt">cn.pdat =</span> <span class="ot">NA</span>, <span class="dt">gep.pdat =</span> <span class="ot">NA</span>)</span></code></pre></div>
</div>
<div id="explanation-of-the-iedge-object-components" class="section level2">
<h2 class="hasAnchor">
<a href="#explanation-of-the-iedge-object-components" class="anchor"></a>Explanation of the iEDGE object components</h2>
<p>The iEDGE object is a list consisting of:</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>cn</strong>: Copy-Number of other binary alteration ExpressionSet:</p>
<ul>
<li><p>exprs(dat$cn): the binary matrix encoding presence or absence of an alteration (c alterations by n samples)</p></li>
<li><p>fData(dat$cn): alteration annotation data frame (c alterations by g alteration annotation columns). Please note that if direction of differential expression test is dependent on the alteration, specify as such to be used in run_iEDGE, e.g. run_iEDGE(… cndir = “direction”, uptest = “UP”, downtest = “DOWN”) corresponds to reporting one-sided pvalues for significantly upregulated in genes for alterations in which fData(dat$cn)[, “direction”] == “UP”, or significantly downregulated genes for alterations in which fData(dat$cn)[, “direction”] == “DOWN”</p></li>
<li><p>pData(dat$cn): sample annotation data frame (n samples by l sample annotation columns)</p></li>
</ul>
</li>
<li>
<p><strong>gep</strong>: The gene expression ExpressionSet:</p>
<ul>
<li><p>exprs(dat$gep): gene expression matrix, assumed to be already log2 transformed (m genes by n samples)</p></li>
<li><p>fData(dat$gep): gene annotation data frame (m genes by j gene annotation columns)</p></li>
<li><p>pData(dat$gep): sample annotation data frame (n samples by k sample annotation columns)</p></li>
</ul>
</li>
<li><p><strong>cisgenes</strong>: A list of character vectors specifying the genes that in the cis regions of each alteration, must be in the same order as the rows in exprs(dat$cn). For copy number alterations we defined the cis genes to be genes in the focal peaks of each amplification or deletion of interest.</p></li>
<li><p><strong>transgenes</strong> (optional): A list of character vectors specifying the genes outside the cis regions (trans genes). If not specified, all genes in the expression matrix that are not the cis genes are considered. Trans genes specification is not recommended for exploratory analysis but may be useful for testing mediation and pathway enrichment for particular target gene sets in hypothesis-driven approaches.</p></li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>dat</span></code></pre></div>
<pre><code>## $cn
## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 2 features, 50 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: sample_1 sample_2 ... sample_50 (50 total)
##   varLabels: colid
##   varMetadata: labelDescription
## featureData
##   featureNames: cn1 cn2
##   fvarLabels: cnid direction
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation:  
## 
## $gep
## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 52 features, 50 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: sample_1 sample_2 ... sample_50 (50 total)
##   varLabels: colid
##   varMetadata: labelDescription
## featureData
##   featureNames: cis1 cis1.1 ... others_30 (52 total)
##   fvarLabels: geneid
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation:  
## 
## $cisgenes
## $cisgenes$cn1
## [1] "cis1"   "cis1.1" "cis1.2"
## 
## $cisgenes$cn2
## [1] "cis2"   "cis2.1" "cis2.2"</code></pre>
</div>
<div id="running-iedge" class="section level2">
<h2 class="hasAnchor">
<a href="#running-iedge" class="anchor"></a>Running iEDGE</h2>
<p>Now that the iEDGE object is created, we can run the iEDGE pipeline on this dataset.</p>
<div id="running-iedge-with-differential-expression-only" class="section level3">
<h3 class="hasAnchor">
<a href="#running-iedge-with-differential-expression-only" class="anchor"></a>Running iEDGE with differential expression only:</h3>
<p>This will create the results object with differential expression tables for cis and trans genes</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># cis and trans differential expression analysis only</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co"># no mediation analysis</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co"># no html report</span></span>
<span id="cb4-4"><a href="#cb4-4"></a>res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/run_iEDGE.html">run_iEDGE</a></span>(<span class="dt">dat =</span> dat, <span class="dt">header =</span> <span class="st">"testrun"</span>, <span class="dt">outdir =</span> <span class="st">"."</span>, </span>
<span id="cb4-5"><a href="#cb4-5"></a>                 <span class="dt">cnid =</span> <span class="st">"cnid"</span>, <span class="dt">cndesc =</span> <span class="st">"cnid"</span>, <span class="dt">cndir =</span> <span class="st">"direction"</span>, </span>
<span id="cb4-6"><a href="#cb4-6"></a>                 <span class="dt">uptest =</span> <span class="st">"UP"</span>, <span class="dt">downtest =</span> <span class="st">"DOWN"</span>,</span>
<span id="cb4-7"><a href="#cb4-7"></a>                 <span class="dt">gepid =</span> <span class="st">"geneid"</span>, <span class="dt">enrichment =</span> <span class="ot">FALSE</span>, <span class="dt">bipartite =</span> <span class="ot">FALSE</span>, <span class="dt">html =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<pre><code>## Running iEDGE for data set: testrun
## Making Differential Expression tables...
## Performancing cis analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_cis_sig_fdr_0.25.txt
## Writing table to ./testrun/tables/testrun_cis_full.txt
## Performancing trans analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_trans_sig_fdr_0.05.txt
## Done!</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(res)</span></code></pre></div>
<pre><code>## [1] "de"        "de.enrich" "pruning"   "ui"</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(res<span class="op">$</span>de<span class="op">$</span>cis<span class="op">$</span>full)</span></code></pre></div>
<pre><code>##        geneid fold.change adj.P.Val.all        t high.class mean1.unlog
## cis1.1 cis1.1    2.132349  3.810924e-24 12.45441       case    15.10178
## cis1.2 cis1.2    2.037804  1.156082e-22 11.70850       case    15.71264
## cis1     cis1    1.916725  3.747522e-20 10.70101       case    15.44336
## cis2     cis2    2.045179  1.156082e-22 11.70744       case    15.49973
## cis2.1 cis2.1    1.918057  3.896869e-20 10.65738       case    15.27832
## cis2.2 cis2.2    1.900450  8.035568e-20 10.50649       case    15.25603
##        mean0.unlog sd0.unlog sd1.unlog cnid direction interaction_type
## cis1.1    6.542553  1.465369  3.168008  cn1        UP              cis
## cis1.2    7.212833  1.903196  3.605576  cn1        UP              cis
## cis1      7.670445  2.081563  3.150070  cn1        UP              cis
## cis2      7.075236  1.886268  3.559967  cn2        UP              cis
## cis2.1    7.446159  1.652364  3.521293  cn2        UP              cis
## cis2.2    7.512069  1.634062  3.505537  cn2        UP              cis
##            logFC  AveExpr      P.Value    adj.P.Val        B mean0.log
## cis1.1 1.0924433 3.634254 6.351540e-25 1.905462e-24 69.81907  2.891393
## cis1.2 1.0270153 3.703061 5.743513e-23 8.615269e-23 60.83393  3.004690
## cis1   0.9386434 3.715198 2.498348e-20 2.498348e-20 49.57859  3.076920
## cis2   1.0322269 3.410606 5.780412e-23 1.734124e-22 60.85680  2.977071
## cis2.1 0.9396453 3.447198 3.247391e-20 4.871087e-20 49.15384  3.052547
## cis2.2 0.9263412 3.452841 8.035568e-20 8.035568e-20 47.56241  3.063777
##        mean1.log   sd0.log   sd1.log n0 n1
## cis1.1  3.983836 0.2661433 0.2704973 16 34
## cis1.2  4.031706 0.3132603 0.3018580 16 34
## cis1    4.015564 0.3486787 0.2625479 16 34
## cis2    4.009298 0.3282196 0.3359340 29 21
## cis2.1  3.992192 0.2774623 0.3171413 29 21
## cis2.2  3.990118 0.2782353 0.3182936 29 21</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(res<span class="op">$</span>de<span class="op">$</span>trans<span class="op">$</span>full)</span></code></pre></div>
<pre><code>##                geneid fold.change adj.P.Val.all        t high.class
## t.m.cis1.1 t.m.cis1.1    2.058719  1.113504e-09 8.257112       case
## t.m.cis1.2 t.m.cis1.2    2.048716  3.095161e-09 7.909354       case
## t.m.cis1     t.m.cis1    1.875885  3.747706e-09 7.773147       case
## t.d.1           t.d.1    2.068440  5.221347e-06 5.636337       case
## t.d.1.1       t.d.1.1    2.036324  7.517011e-06 5.514853       case
## t.d.1.2       t.d.1.2    1.910407  5.188124e-04 4.259915       case
##            mean1.unlog mean0.unlog sd0.unlog sd1.unlog cnid direction
## t.m.cis1.1   130.97073   62.936666 22.403932 43.210031  cn1        UP
## t.m.cis1.2   137.37791   68.380230 26.189159 37.396372  cn1        UP
## t.m.cis1     131.10487   70.856702 21.898342 30.521851  cn1        UP
## t.d.1         15.96287    7.664441  5.865252  7.642273  cn1        UP
## t.d.1.1       16.21092    7.938313  5.287850  6.876634  cn1        UP
## t.d.1.2       16.02314    8.130853  5.495517  8.454433  cn1        UP
##            interaction_type     logFC  AveExpr      P.Value    adj.P.Val
## t.m.cis1.1            trans 1.0417471 6.645719 4.544915e-11 2.227009e-09
## t.m.cis1.2            trans 1.0347199 6.729740 1.623003e-10 3.976358e-09
## t.m.cis1              trans 0.9075714 6.716427 2.676933e-10 4.372324e-09
## t.d.1                 trans 1.0485433 3.626562 6.926277e-07 8.484689e-06
## t.d.1.1               trans 1.0259674 3.666232 1.073859e-06 1.052382e-05
## t.d.1.2               trans 0.9338801 3.618542 8.470406e-05 6.917498e-04
##                     B mean0.log mean1.log   sd0.log   sd1.log n0 n1
## t.m.cis1.1 14.9623347  5.937331  6.979078 0.4059199 0.4280755 16 34
## t.m.cis1.2 13.7029102  6.026131  7.060851 0.5216744 0.3940067 16 34
## t.m.cis1   13.2079775  6.099279  7.006850 0.4695779 0.3438308 16 34
## t.d.1       5.4640829  2.913553  3.962096 0.7221742 0.5880397 16 34
## t.d.1.1     5.0346486  2.968574  3.994541 0.7366806 0.5798271 16 34
## t.d.1.2     0.7893726  2.983503  3.917383 0.7726423 0.7383662 16 34</code></pre>
</div>
<div id="running-iedge-with-differential-expression-only-and-html-report" class="section level3">
<h3 class="hasAnchor">
<a href="#running-iedge-with-differential-expression-only-and-html-report" class="anchor"></a>Running iEDGE with differential expression only and HTML report:</h3>
<p>This will create the results object with differential expression tables for cis and trans genes only, and an associated HTML report</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co"># cis and trans differential expression analysis only</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="co"># no mediation analysis</span></span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="co"># do html report</span></span>
<span id="cb12-4"><a href="#cb12-4"></a>res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/run_iEDGE.html">run_iEDGE</a></span>(<span class="dt">dat =</span> dat, <span class="dt">header =</span> <span class="st">"testrun"</span>, <span class="dt">outdir =</span> <span class="st">"."</span>, </span>
<span id="cb12-5"><a href="#cb12-5"></a>                 <span class="dt">cnid =</span> <span class="st">"cnid"</span>, <span class="dt">cndesc =</span> <span class="st">"cnid"</span>, <span class="dt">cndir =</span> <span class="st">"direction"</span>, </span>
<span id="cb12-6"><a href="#cb12-6"></a>                 <span class="dt">uptest =</span> <span class="st">"UP"</span>, <span class="dt">downtest =</span> <span class="st">"DOWN"</span>,</span>
<span id="cb12-7"><a href="#cb12-7"></a>                 <span class="dt">gepid =</span> <span class="st">"geneid"</span>, <span class="dt">enrichment =</span> <span class="ot">FALSE</span>, </span>
<span id="cb12-8"><a href="#cb12-8"></a>                 <span class="dt">bipartite =</span> <span class="ot">FALSE</span>, <span class="dt">html =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## Running iEDGE for data set: testrun
## Making Differential Expression tables...
## Performancing cis analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_cis_sig_fdr_0.25.txt
## Writing table to ./testrun/tables/testrun_cis_full.txt
## Performancing trans analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_trans_sig_fdr_0.05.txt
## Done!
## Writing by alteration cis full DE tables...
## Writing cis boxplots...
## Writing trans boxplots...
## Writing by alteration cis DE tables...
## Writing by alteration trans DE tables...
## Writing summary table...
## Adding links to main table...
## Adding main table to main index...
## Done constructing html report at directory: ./testrun/html</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co"># the result html page should be opened automatically, this displays the graphical view of results</span></span></code></pre></div>
</div>
<div id="running-iedge-with-differential-expression-and-mediation-analysis-bipartite-and-html-report" class="section level3">
<h3 class="hasAnchor">
<a href="#running-iedge-with-differential-expression-and-mediation-analysis-bipartite-and-html-report" class="anchor"></a>Running iEDGE with differential expression and mediation analysis (bipartite) and HTML report:</h3>
<p>This will create the results object with differential expression tables for cis and trans genes <em>and</em> mediation analysis, and an associated HTML report</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="co"># cis and trans differential expression analysis only</span></span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="co"># do mediation analysis</span></span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="co"># do html report</span></span>
<span id="cb15-4"><a href="#cb15-4"></a>res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/run_iEDGE.html">run_iEDGE</a></span>(<span class="dt">dat =</span> dat, <span class="dt">header =</span> <span class="st">"testrun"</span>, <span class="dt">outdir =</span> <span class="st">"."</span>, </span>
<span id="cb15-5"><a href="#cb15-5"></a>                 <span class="dt">cnid =</span> <span class="st">"cnid"</span>, <span class="dt">cndesc =</span> <span class="st">"cnid"</span>, <span class="dt">cndir =</span> <span class="st">"direction"</span>, </span>
<span id="cb15-6"><a href="#cb15-6"></a>                 <span class="dt">uptest =</span> <span class="st">"UP"</span>, <span class="dt">downtest =</span> <span class="st">"DOWN"</span>,</span>
<span id="cb15-7"><a href="#cb15-7"></a>                 <span class="dt">gepid =</span> <span class="st">"geneid"</span>, <span class="dt">enrichment =</span> <span class="ot">FALSE</span>, </span>
<span id="cb15-8"><a href="#cb15-8"></a>                 <span class="dt">bipartite =</span> <span class="ot">TRUE</span>, <span class="dt">html =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## Running iEDGE for data set: testrun
## Making Differential Expression tables...
## Performancing cis analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_cis_sig_fdr_0.25.txt
## Writing table to ./testrun/tables/testrun_cis_full.txt
## Performancing trans analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_trans_sig_fdr_0.05.txt
## Done!
## Running pruning for  cn1 
## [1] 1
## [1] "number cis"
## [1] 3
## [1] "number trans"
## [1] 6
## Performing cis ordering...
## Performing trans ordering on 5 sets...
## 
## Running pruning for  cn2 
## [1] 2
## [1] "number cis"
## [1] 3
## [1] "number trans"
## [1] 10
## Performing cis ordering...
## Performing trans ordering on 8 sets...
## 
## Writing by alteration cis full DE tables...
## Writing cis boxplots...
## Writing trans boxplots...
## Writing by alteration cis DE tables...
## Writing by alteration trans DE tables...
## Writing summary table...
## Adding links to main table...
## Adding main table to main index...
## Writing bipartite graphs...
## Done constructing html report at directory: ./testrun/html</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="co"># significant cn-cis-trans connections found using mediation analysis</span></span>
<span id="cb17-2"><a href="#cb17-2"></a>res<span class="op">$</span>pruning<span class="op">$</span>sig</span></code></pre></div>
<pre><code>## $cn1
##      cis      trans yind zind   taudiff      tau0         tau1  tauratio
## 1 cis1.2      t.d.1    2    4 0.5420271 1.0485433  0.506516245 0.5169334
## 3 cis1.1    t.d.1.2    1    6 1.2585709 0.9338801 -0.324690839 1.3476794
## 4   cis1   t.m.cis1    3    3 0.7631243 0.9075714  0.144447104 0.8408422
## 5 cis1.1 t.m.cis1.1    1    1 1.2319340 1.0417471 -0.190186947 1.1825654
## 6   cis1 t.m.cis1.2    3    2 1.0403221 1.0347199 -0.005602134 1.0054142
##          Z         S       pvalue          fdr
## 1 1.782062 0.3041573 3.736959e-02 1.345305e-01
## 3 3.035446 0.4146247 1.200901e-03 5.404054e-03
## 4 4.753845 0.1605278 9.979214e-07 5.987529e-06
## 5 6.287764 0.1959256 1.610360e-10 2.898648e-09
## 6 6.139045 0.1694599 4.150955e-10 3.735860e-09
## 
## $cn2
##       cis        trans yind zind   taudiff      tau0        tau1  tauratio
## 3  cis2.1      t.d.2.2    2    8 0.4916785 1.0756830  0.58400453 0.4570849
## 4    cis2     t.m.cis2    1    1 0.8209718 1.1270505  0.30607867 0.7284251
## 5  cis2.1   t.m.cis2.1    2    3 0.8365285 0.9352537  0.09872523 0.8944402
## 6  cis2.1 t.m.cis2.1.1    2    6 1.0545167 0.8939961 -0.16052060 1.1795540
## 7  cis2.1 t.m.cis2.1.2    2    7 0.9844624 0.8363706 -0.14809181 1.1770648
## 8  cis2.2   t.m.cis2.2    3    2 0.8995824 0.9720952  0.07251276 0.9254057
## 9    cis2   t.m.cis2.3    1    5 0.9507412 0.9825324  0.03179122 0.9676436
## 10   cis2   t.m.cis2.4    1    4 1.2855678 1.1221337 -0.16343419 1.1456459
##           Z         S       pvalue          fdr
## 3  1.758017 0.2796779 3.937231e-02 1.476461e-01
## 4  4.995617 0.1643384 2.932394e-07 1.256740e-06
## 5  5.773441 0.1448925 3.883433e-09 2.912574e-08
## 6  6.737861 0.1565061 8.036751e-12 1.205513e-10
## 7  5.507752 0.1787412 1.817228e-08 1.090337e-07
## 8  6.483173 0.1387565 4.490663e-11 4.490663e-10
## 9  5.116760 0.1858092 1.554143e-07 7.770716e-07
## 10 7.032506 0.1828037 1.014282e-12 3.042846e-11</code></pre>
<p>The result html page should open automatically, and it will include an additional column “bipartite” with hyperlinks to the graphical display of the mediation analysis.</p>
</div>
</div>
<div id="iedge-html-report" class="section level2">
<h2 class="hasAnchor">
<a href="#iedge-html-report" class="anchor"></a>iEDGE HTML report</h2>
<p>The HTML report of the final run is contained <a href="./testrun/html/index.html">here</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#setting-up-the-iedge-object">Setting up the iEDGE object</a></li>
      <li><a href="#explanation-of-the-iedge-object-components">Explanation of the iEDGE object components</a></li>
      <li><a href="#running-iedge">Running iEDGE</a></li>
      <li><a href="#iedge-html-report">iEDGE HTML report</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Amy Li, Stefano Monti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
