<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>iEDGE - tutorial for simple run using simulated dataset â€¢ iEDGE</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="iEDGE - tutorial for simple run using simulated dataset">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">iEDGE</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/iedge-simple-run.html">iEDGE - tutorial for simple run using simulated dataset</a>
    </li>
    <li>
      <a href="../articles/iedge-tcga-example.html">iEDGE - tutorial for example walkthrough on a subset of TCGA PAAD dataset</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/montilab/iEDGE">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>iEDGE - tutorial for simple run using simulated dataset</h1>
                        <h4 class="author">Amy Li</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/montilab/iEDGE/blob/master/vignettes/iedge-simple-run.Rmd"><code>vignettes/iedge-simple-run.Rmd</code></a></small>
      <div class="hidden name"><code>iedge-simple-run.Rmd</code></div>

    </div>

    
    
<div id="getting-started-with-iedge-walkthrough-with-simple-simulated-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started-with-iedge-walkthrough-with-simple-simulated-dataset" class="anchor"></a>Getting started with iEDGE: walkthrough with simple simulated dataset</h1>
<div id="setting-up-the-iedge-object" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-iedge-object" class="anchor"></a>Setting up the iEDGE object</h2>
<p>For this tutorial we will simulate the input gene expression data, the cis and trans genes with known mediations and run iEDGE to recover these mediations.</p>
<p>We start by generating the dataset. You can use construct_iEDGE to generate the iEDGE dataset from standard matrices (binary matrices for alteration data or numeric matrices for expression data).</p>
<p>Please note that this simulation is optimized for the evaluation of the iEDGE methology but is meant for demonstrating how to construct an iEDGE object from scratch for use in standard analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(iEDGE)</code></pre></div>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: 'BiocGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, cbind, colMeans,
##     colnames, colSums, do.call, duplicated, eval, evalq, Filter,
##     Find, get, grep, grepl, intersect, is.unsorted, lapply,
##     lengths, Map, mapply, match, mget, order, paste, pmax,
##     pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce,
##     rowMeans, rownames, rowSums, sapply, setdiff, sort, table,
##     tapply, union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)

<span class="co">#define number of samples</span>
n&lt;-<span class="dv">50</span>

<span class="co">#construct 'cn': binary alternations by samples matrix </span>
<span class="co">#two independent alterations across 50 samples</span>
cn1&lt;-<span class="kw">sample</span>(<span class="dt">x=</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">0</span>), <span class="dt">size =</span> n, <span class="dt">replace =</span> T, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="fl">0.7</span>, <span class="fl">0.3</span>))
cn2&lt;-<span class="kw">sample</span>(<span class="dt">x=</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">0</span>), <span class="dt">size =</span> n, <span class="dt">replace =</span> T, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>))

cn&lt;-<span class="kw">rbind</span>(cn1, cn2)
<span class="kw">colnames</span>(cn)&lt;-<span class="kw">paste0</span>(<span class="st">"sample_"</span>, <span class="dv">1</span>:<span class="kw">ncol</span>(cn))

<span class="co">#construct 'cis' gene expressions: genes correlated with one or more alterations in 'cn'</span>
cis&lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">sapply</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>), function(i){
        <span class="dv">3</span> +<span class="st"> </span>cn[i, ] +<span class="st"> </span><span class="kw">rnorm</span>(n, <span class="dv">0</span>, <span class="fl">0.3</span>)
    }))
<span class="kw">rownames</span>(cis)&lt;-<span class="kw">make.unique</span>(<span class="kw">paste0</span>(<span class="st">"cis"</span>, <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>)))

<span class="co">#construct 'trans-mediated' gene expression: gene correlated with one or more alteration in 'cn' and mediated through a cis gene</span>
cis.m&lt;-<span class="kw">sample</span>(<span class="kw">rownames</span>(cis), <span class="dv">10</span>, <span class="dt">replace =</span> T)
trans.m&lt;-<span class="kw">t</span>(<span class="kw">sapply</span>(<span class="dv">1</span>:<span class="kw">length</span>(cis.m), function(i)
        <span class="dv">3</span> +<span class="st"> </span>cis[cis.m[i], ] +<span class="st"> </span><span class="kw">rnorm</span>(n, <span class="dv">0</span>, <span class="fl">0.3</span>)
    ))
<span class="kw">rownames</span>(trans.m)&lt;-<span class="kw">make.unique</span>(<span class="kw">paste0</span>(<span class="st">"t.m."</span>, cis.m))

<span class="co">#construct 'trans-direct' gene expression: trans gene correlated with one or more alteration in 'cn' but not mediated through cis genes, this would be similar to construction of cis genes</span>
trans.d&lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">sapply</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>), function(i)
        <span class="dv">3</span> +<span class="st"> </span>cn[i, ] +<span class="st"> </span><span class="kw">rnorm</span>(n, <span class="dv">0</span>, <span class="fl">0.6</span>)
        ))
<span class="kw">rownames</span>(trans.d)&lt;-<span class="kw">make.unique</span>(<span class="kw">paste0</span>(<span class="st">"t.d."</span>, <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>)))


<span class="co">#uncorrelated genes</span>
others&lt;-<span class="kw">t</span>(<span class="kw">sapply</span>(<span class="dv">1</span>:<span class="dv">30</span>, function(i) <span class="dv">3</span>+<span class="kw">rnorm</span>(n, <span class="dv">0</span>, <span class="fl">0.3</span>)))
<span class="kw">rownames</span>(others)&lt;-<span class="kw">paste0</span>(<span class="st">"others_"</span>, <span class="dv">1</span>:<span class="kw">nrow</span>(others))

<span class="co">#putting genes together in expression matrix</span>
gep&lt;-<span class="kw">rbind</span>(cis, trans.m, trans.d, others)
<span class="kw">colnames</span>(gep)&lt;-<span class="kw">paste0</span>(<span class="st">"sample_"</span>, <span class="dv">1</span>:<span class="kw">ncol</span>(gep))

cisgenes&lt;-<span class="kw">list</span>(<span class="dt">cn1 =</span> <span class="kw">grep</span>(<span class="st">"cis1"</span>, <span class="kw">rownames</span>(cis), <span class="dt">value =</span> T),
               <span class="dt">cn2 =</span> <span class="kw">grep</span>(<span class="st">"cis2"</span>, <span class="kw">rownames</span>(cis), <span class="dt">value =</span> T))

<span class="co">#the direction column in cn.fdat specifies one-sided vs. bidirectional p-values in differential test</span>
<span class="co">#in this simulation we expect increase in expression with the presence of the alteration</span>
cn.fdat&lt;-<span class="kw">data.frame</span>(<span class="dt">cnid =</span> <span class="kw">rownames</span>(cn), <span class="dt">direction =</span> <span class="st">"UP"</span>)
<span class="kw">rownames</span>(cn.fdat)&lt;-<span class="kw">rownames</span>(cn)

gep.fdat&lt;-<span class="kw">data.frame</span>(<span class="dt">geneid =</span> <span class="kw">rownames</span>(gep))
<span class="kw">rownames</span>(gep.fdat)&lt;-<span class="kw">rownames</span>(gep)


dat&lt;-<span class="kw"><a href="../reference/construct_iEDGE.html">construct_iEDGE</a></span>(cn, gep, cisgenes, <span class="dt">transgenes =</span> <span class="ot">NA</span>, cn.fdat, gep.fdat, 
                <span class="dt">cn.pdat =</span> <span class="ot">NA</span>, <span class="dt">gep.pdat =</span> <span class="ot">NA</span>)</code></pre></div>
<pre><code>## Warning in to.eSet(mat = cn, pdat = cn.pdat, fdat = cn.fdat): pdat rownames
## and mat colnames do not match, setting fdat rownames to mat rownames</code></pre>
<pre><code>## Warning in to.eSet(mat = gep, pdat = gep.pdat, fdat = gep.fdat): pdat
## rownames and mat colnames do not match, setting fdat rownames to mat
## rownames</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat</code></pre></div>
<pre><code>## $cn
## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 2 features, 50 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: sample_1 sample_2 ... sample_50 (50 total)
##   varLabels: colid
##   varMetadata: labelDescription
## featureData
##   featureNames: cn1 cn2
##   fvarLabels: cnid direction
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation:  
## 
## $gep
## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 52 features, 50 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: sample_1 sample_2 ... sample_50 (50 total)
##   varLabels: colid
##   varMetadata: labelDescription
## featureData
##   featureNames: cis1 cis1.1 ... others_30 (52 total)
##   fvarLabels: geneid
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation:  
## 
## $cisgenes
## $cisgenes$cn1
## [1] "cis1"   "cis1.1" "cis1.2"
## 
## $cisgenes$cn2
## [1] "cis2"   "cis2.1" "cis2.2"</code></pre>
</div>
<div id="running-iedge" class="section level2">
<h2 class="hasAnchor">
<a href="#running-iedge" class="anchor"></a>Running iEDGE</h2>
<p>Now that the iEDGE object is created, we can run the iEDGE pipeline on this dataset</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#cis and trans differential expression analysis only</span>
<span class="co">#no mediation analysis</span>
<span class="co">#no html report</span>
res&lt;-<span class="kw"><a href="../reference/run_iEDGE.html">run_iEDGE</a></span>(<span class="dt">dat =</span> dat, <span class="dt">header =</span> <span class="st">"testrun"</span>, <span class="dt">outdir =</span> <span class="st">"."</span>, 
        <span class="dt">cnid =</span> <span class="st">"cnid"</span>, <span class="dt">cndesc =</span> <span class="st">"cnid"</span>, <span class="dt">cndir =</span> <span class="st">"direction"</span>, 
        <span class="dt">uptest =</span> <span class="st">"UP"</span>, <span class="dt">downtest =</span> <span class="st">"DOWN"</span>,
        <span class="dt">gepid =</span> <span class="st">"geneid"</span>, <span class="dt">enrichment =</span> <span class="ot">FALSE</span>, <span class="dt">bipartite =</span> <span class="ot">FALSE</span>, <span class="dt">html =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## Running iEDGE for data set: testrun
## Making Differential Expression tables...
## Performancing cis analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_cis_sig_fdr_0.25.txt
## Writing table to ./testrun/tables/testrun_cis_full.txt
## Performancing trans analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_trans_sig_fdr_0.05.txt
## Done!</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#cis and trans differential expression analysis only</span>
<span class="co">#no mediation analysis</span>
<span class="co">#do html report</span>
res&lt;-<span class="kw"><a href="../reference/run_iEDGE.html">run_iEDGE</a></span>(<span class="dt">dat =</span> dat, <span class="dt">header =</span> <span class="st">"testrun"</span>, <span class="dt">outdir =</span> <span class="st">"."</span>, 
        <span class="dt">cnid =</span> <span class="st">"cnid"</span>, <span class="dt">cndesc =</span> <span class="st">"cnid"</span>, <span class="dt">cndir =</span> <span class="st">"direction"</span>, 
        <span class="dt">uptest =</span> <span class="st">"UP"</span>, <span class="dt">downtest =</span> <span class="st">"DOWN"</span>,
        <span class="dt">gepid =</span> <span class="st">"geneid"</span>, <span class="dt">enrichment =</span> <span class="ot">FALSE</span>, <span class="dt">bipartite =</span> <span class="ot">FALSE</span>, <span class="dt">html =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## Running iEDGE for data set: testrun
## Making Differential Expression tables...
## Performancing cis analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_cis_sig_fdr_0.25.txt
## Writing table to ./testrun/tables/testrun_cis_full.txt
## Performancing trans analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_trans_sig_fdr_0.05.txt
## Done!
## Writing by alteration cis full DE tables...
## Writing cis boxplots...
## Writing trans boxplots...
## Writing by alteration cis DE tables...
## Writing by alteration trans DE tables...
## Writing summary table...
## Adding links to main table...
## Adding main table to main index...
## Done constructing html report at directory: ./testrun/html</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#cis and trans differential expression analysis only</span>
<span class="co">#do mediation analysis</span>
<span class="co">#do html report</span>
res&lt;-<span class="kw"><a href="../reference/run_iEDGE.html">run_iEDGE</a></span>(<span class="dt">dat =</span> dat, <span class="dt">header =</span> <span class="st">"testrun"</span>, <span class="dt">outdir =</span> <span class="st">"."</span>, 
        <span class="dt">cnid =</span> <span class="st">"cnid"</span>, <span class="dt">cndesc =</span> <span class="st">"cnid"</span>, <span class="dt">cndir =</span> <span class="st">"direction"</span>, 
        <span class="dt">uptest =</span> <span class="st">"UP"</span>, <span class="dt">downtest =</span> <span class="st">"DOWN"</span>,
        <span class="dt">gepid =</span> <span class="st">"geneid"</span>, <span class="dt">enrichment =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## Running iEDGE for data set: testrun
## Making Differential Expression tables...
## Performancing cis analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_cis_sig_fdr_0.25.txt
## Writing table to ./testrun/tables/testrun_cis_full.txt
## Performancing trans analysis...
## Running iEDGE differential expression for 2 alterations: 
## 
## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |================================                                 |  50%
  |                                                                       
  |=================================================================| 100%
## Writing table to ./testrun/tables/testrun_trans_sig_fdr_0.05.txt
## Done!
## Running pruning for  cn1 
## [1] 1
## [1] "number cis"
## [1] 3
## [1] "number trans"
## [1] 5
## Performing cis ordering...
## Performing trans ordering on 2 sets...
## 
## Running pruning for  cn2 
## [1] 2
## [1] "number cis"
## [1] 3
## [1] "number trans"
## [1] 14
## Performing cis ordering...
## Performing trans ordering on 8 sets...
## 
## Writing by alteration cis full DE tables...
## Writing cis boxplots...
## Writing trans boxplots...
## Writing by alteration cis DE tables...
## Writing by alteration trans DE tables...
## Writing summary table...
## Adding links to main table...
## Adding main table to main index...
## Writing bipartite graphs...
## Done constructing html report at directory: ./testrun/html</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#the result html page should be opened automatically, this displays the graphical view of results</span>
<span class="co">#you can also view the results in raw matrix format</span>
<span class="kw">names</span>(res)</code></pre></div>
<pre><code>## [1] "de"        "de.enrich" "pruning"   "ui"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(res$de$cis$sig)</code></pre></div>
<pre><code>##        geneid fold.change adj.P.Val.all        t high.class mean1.unlog
## cis1.1 cis1.1    2.132349  3.810924e-24 12.45441       case    15.10178
## cis1.2 cis1.2    2.037804  1.156082e-22 11.70850       case    15.71264
## cis1     cis1    1.916725  3.747522e-20 10.70101       case    15.44336
## cis2     cis2    2.045179  1.156082e-22 11.70744       case    15.49973
## cis2.1 cis2.1    1.918057  3.896869e-20 10.65738       case    15.27832
## cis2.2 cis2.2    1.900450  8.035568e-20 10.50649       case    15.25603
##        mean0.unlog sd0.unlog sd1.unlog cnid direction interaction_type
## cis1.1    6.542553  1.465369  3.168008  cn1        UP              cis
## cis1.2    7.212833  1.903196  3.605576  cn1        UP              cis
## cis1      7.670445  2.081563  3.150070  cn1        UP              cis
## cis2      7.075236  1.886268  3.559967  cn2        UP              cis
## cis2.1    7.446159  1.652364  3.521293  cn2        UP              cis
## cis2.2    7.512069  1.634062  3.505537  cn2        UP              cis
##            logFC  AveExpr      P.Value    adj.P.Val        B mean0.log
## cis1.1 1.0924433 3.634254 6.351540e-25 1.905462e-24 69.81907  2.891393
## cis1.2 1.0270153 3.703061 5.743513e-23 8.615269e-23 60.83393  3.004690
## cis1   0.9386434 3.715198 2.498348e-20 2.498348e-20 49.57859  3.076920
## cis2   1.0322269 3.410606 5.780412e-23 1.734124e-22 60.85680  2.977071
## cis2.1 0.9396453 3.447198 3.247391e-20 4.871087e-20 49.15384  3.052547
## cis2.2 0.9263412 3.452841 8.035568e-20 8.035568e-20 47.56241  3.063777
##        mean1.log   sd0.log   sd1.log n0 n1
## cis1.1  3.983836 0.2661433 0.2704973 16 34
## cis1.2  4.031706 0.3132603 0.3018580 16 34
## cis1    4.015564 0.3486787 0.2625479 16 34
## cis2    4.009298 0.3282196 0.3359340 29 21
## cis2.1  3.992192 0.2774623 0.3171413 29 21
## cis2.2  3.990118 0.2782353 0.3182936 29 21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(res$de$trans$sig)</code></pre></div>
<pre><code>##                geneid fold.change adj.P.Val.all         t high.class
## t.m.cis1.2 t.m.cis1.2    1.872049  2.453809e-08  7.224216       case
## t.m.cis1     t.m.cis1    1.928035  6.777171e-08  6.863515       case
## t.d.1           t.d.1    2.085097  1.580501e-07  6.610930       case
## t.d.1.2       t.d.1.2    2.099057  5.983020e-06  5.557481       case
## t.d.1.1       t.d.1.1    1.758332  1.603663e-04  4.606910       case
## t.m.cis21    t.m.cis2    2.099416  2.654605e-12 10.238436       case
##            mean1.unlog mean0.unlog sd0.unlog sd1.unlog cnid direction
## t.m.cis1.2   125.47707   65.783097 17.256715 40.441390  cn1        UP
## t.m.cis1     128.56393   68.256628 26.732052 37.636026  cn1        UP
## t.d.1         16.46298    7.084557  2.475993  7.196020  cn1        UP
## t.d.1.2       16.07174    6.938782  2.961621  7.847104  cn1        UP
## t.d.1.1       16.78763    9.636795  5.216208  6.835061  cn1        UP
## t.m.cis21    133.62851   63.350522 15.799246 30.256475  cn2        UP
##            interaction_type     logFC  AveExpr      P.Value    adj.P.Val
## t.m.cis1.2            trans 0.9046180 6.628850 2.003110e-09 9.815238e-08
## t.m.cis1              trans 0.9471311 6.657309 7.607029e-09 1.863722e-07
## t.d.1                 trans 1.0601142 3.672923 1.935307e-08 3.161002e-07
## t.d.1.2               trans 1.0697413 3.605716 9.157683e-07 1.121816e-05
## t.d.1.1               trans 0.8142077 3.797642 2.618225e-05 2.565861e-04
## t.m.cis21             trans 1.0699878 6.412925 2.708781e-14 1.327303e-12
##                    B mean0.log mean1.log   sd0.log   sd1.log n0 n1
## t.m.cis1.2 11.296754  6.013710  6.918328 0.3890997 0.4299934 16 34
## t.m.cis1    9.984859  6.013260  6.960391 0.5627077 0.4102757 16 34
## t.d.1       9.067678  2.952045  4.012159 0.4433943 0.5822797 16 34
## t.d.1.2     5.291036  2.878292  3.948033 0.6211865 0.6711068 16 34
## t.d.1.1     2.034074  3.243981  4.058189 0.7392263 0.5254953 16 34
## t.m.cis21  22.205987  5.963531  7.033518 0.3717169 0.3589295 29 21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#significant cn-cis-trans connections found using mediation analysis</span>
res$pruning$sig</code></pre></div>
<pre><code>## $cn1
##      cis      trans yind zind   taudiff      tau0         tau1  tauratio
## 7   cis1   t.m.cis1    3    2 1.0332525 0.9471311 -0.086121413 1.0909287
## 8 cis1.2 t.m.cis1.2    2    1 0.8975216 0.9046180  0.007096426 0.9921553
##          Z         S       pvalue          fdr
## 7 5.609721 0.1841897 1.013268e-08 1.519901e-07
## 8 5.068113 0.1770919 2.008894e-07 1.506670e-06
## 
## $cn2
##       cis        trans yind zind   taudiff      tau0         tau1
## 9    cis2     t.m.cis2    1    1 0.7181678 1.0699878  0.351819996
## 10 cis2.1   t.m.cis2.1    2    4 1.0898218 1.0804565 -0.009365314
## 11 cis2.1 t.m.cis2.1.1    2    5 1.0448972 0.9626621 -0.082235112
## 12 cis2.1 t.m.cis2.1.2    2   10 0.7975211 0.8152441  0.017723002
## 13 cis2.2   t.m.cis2.2    3    6 0.8649442 0.8672242  0.002279970
## 14 cis2.2 t.m.cis2.2.1    3    3 0.9803627 1.0122671  0.031904455
## 15 cis2.2 t.m.cis2.2.2    3    2 0.8258147 0.9629575  0.137142825
## 16   cis2   t.m.cis2.3    1    7 0.9577238 0.9496609 -0.008062906
##     tauratio        Z         S       pvalue          fdr
## 9  0.6711925 4.943945 0.1452621 3.827866e-07 2.296719e-06
## 10 1.0086679 5.663790 0.1924192 7.403266e-09 1.295405e-07
## 11 1.0854247 6.424853 0.1626336 6.599871e-11 2.771946e-09
## 12 0.9782605 4.280637 0.1863090 9.317944e-06 4.891921e-05
## 13 0.9973710 5.421178 0.1595491 2.960384e-08 3.108403e-07
## 14 0.9684822 5.625419 0.1742737 9.252891e-09 1.295405e-07
## 15 0.8575816 5.076170 0.1626846 1.925591e-07 1.347913e-06
## 16 1.0084903 5.299893 0.1807063 5.793537e-08 4.866571e-07</code></pre>
</div>
<div id="iedge-html-report" class="section level2">
<h2 class="hasAnchor">
<a href="#iedge-html-report" class="anchor"></a>iEDGE HTML report</h2>
<p>The HTML report of this run is contained <a href="file:./testrun/html/index.html">here</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#getting-started-with-iedge-walkthrough-with-simple-simulated-dataset">Getting started with iEDGE: walkthrough with simple simulated dataset</a><ul class="nav nav-pills nav-stacked">
<li><a href="#setting-up-the-iedge-object">Setting up the iEDGE object</a></li>
      <li><a href="#running-iedge">Running iEDGE</a></li>
      <li><a href="#iedge-html-report">iEDGE HTML report</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Amy Li.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
